lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb

```diff
@@ -0,0 +1,202 @@
+require "topological_inventory/ansible_tower/collector"
+require "topological_inventory/providers/common/operations/sources_api_client"
+require "topological_inventory/ansible_tower/receptor/async_receiver"
+
+module TopologicalInventory
+  module AnsibleTower
+    module TargetedRefresh
+      class ServiceInstance < TopologicalInventory::AnsibleTower::Collector
+        REFS_PER_REQUEST_LIMIT = 20
+
+        def initialize(payload = {})
+          self.params    = payload['params']
+          self.source_id = payload['source_id']
+
+          # TODO: add metrics exporter
+          super(payload['source_uid'], nil)
+        end
+
+        # Entrypoint for 'ServiceInstance.refresh' operation
+        def refresh
+          return if params_missing?
+
+          # Connection settings
+          set_connection_data!
+          logger.info("ServiceInstance#refresh - Connection set successfully")
+
+          # Get all tasks
+          tasks, refresh_state_uuid = {}, SecureRandom.uuid
+          params.to_a.each do |task|
+            if task['task_id'].blank? || task['source_ref'].blank?
+              logger.warn("ServiceInstance#refresh - missing data for task: #{task}")
+              next
+            end
+            tasks[task['task_id']] = task['source_ref']
+
+            if tasks.length == REFS_PER_REQUEST_LIMIT
+              refresh_part(tasks, refresh_state_uuid, SecureRandom.uuid)
+              tasks = {}
+            end
+          end
+
+          refresh_part(tasks, refresh_state_uuid, SecureRandom.uuid) unless tasks.empty?
+        rescue => err
+          logger.error("ServiceInstance#refresh - Error: #{err.message}\n#{err.backtrace.join("\n")}")
+        end
+
+        private
+
+        attr_accessor :params, :source_id
+
+        def required_params
+          %w[source_id source params]
+        end
+
+        def params_missing?
+          is_missing = false
+          required_params.each do |attr|
+            if (is_missing = send(attr).blank?)
```

Comment:
Is the `if (is_missing = send(attr).blank?)` pattern used for readability?  I'm not sure if  `if(send(attr).blank?)` would be harder to read since we're iterating `required_params`.

Reply:
@syncrou this method is duplicated accross operations, so it'll be moved to providers-common in next refactoring. 
But I'm not sure what you are suggesting

Reply:
I don't see the need for the extra boolean:
```
          required_params.each do |attr|
            if (send(attr).blank?)
```

Probably worth a discussion later. Nothing to hold up this PR.